doc/spire_agent.md

```diff
@@ -76,10 +76,60 @@ Please see the [built-in plugins](#built-in-plugins) section for information on
 
 ## Telemetry configuration
 
-If telemetry is desired, it may be configured by using a dedicated `telemetry { ... }` section. Prometheus, Statsd, and DogStatsd are currently supported. You may use all, some, or none. Statsd and DogStatsd both support multiple declarations in the event that you want to send metrics to more than one collector. Here is a sample configuration:
+If telemetry is desired, it may be configured by using a dedicated `telemetry { ... }` section. The following metrics collectors are currently supported:
+- Prometheus
+- Statsd
+- DogStatsd
+- M3
+
+You may use all, some, or none. The following collectors support multiple declarations in the event that you want to send metrics to more than one collector:
+
+- Statsd
+- DogStatsd
+- M3
+
+### Telemetry configuration syntax
+
+| Configuration          | Type          | Description  | Default |
+| ----------------       | ------------- | ------------ | ------- |
+| `EnableTypePrefix`     | `bool`        | Includes metric type information as the second key following the service name | false |
+| `Prometheus`           | `Prometheus`  | Prometheus configuration         | |
+| `DogStatsd`            | `[]DogStatsd` | List of DogStatsd configurations | |
+| `Statsd`               | `[]Statsd`    | List of Statsd configurations    | |
+| `M3`                   | `[]M3`        | List of M3 configurations        | |
+
+#### `Prometheus`
+
+| Configuration    | Type          | Description |
+| ---------------- | ------------- | ----------- |
+| `host`           | `string`      | Prometheus server host |
+| `port`           | `int`         | Prometheus server port |
+
+#### `DogStatsd`
+| Configuration    | Type          | Description |
+| ---------------- | ------------- | ----------- |
+| `address`        | `string`      | DogStatsd address |
+
+#### `Statsd`
+| Configuration    | Type          | Description |
+| ---------------- | ------------- | ----------- |
+| `address`        | `string`      | Statsd address |
+
+#### `M3`
+| Configuration    | Type          | Description |
+| ---------------- | ------------- | ----------- |
+| `address`        | `string`      | M3 address |
+| `env`            | `string`      | M3 environment, e.g. `production`, `staging` |
+
+When M3 collectors are used, it is highly recommended to set `EnableTypePrefix` to `true` in order to ensure all metrics are emitted properly.
```

Comment:
When does it make sense not to set EnableTypePrefix to true for M3? Would it be better to imply EnableTypePrefix if an M3 collector is set?

Reply:
IMO it would feel odd to have the existence of an implementation (m3) force a config change on the abstraction (go-metrics).
Also, I recall there being a discussion about whether it still makes sense to allow/suggest setting up of multiple metrics configurations at once. However if we're still in the mode of allowing as many as the user wants, they would need to accept that the existence of m3 forces a metric prefix onto all of their sinks, rather than allowing a choice of having the other sinks run as-is and go against m3 recommendation (for whatever reason).

Reply:
I added the `EnableTypePrefix` prefix for the reasons that @amoore877 mentioned above. It feels clunky to require this additional config parameter for M3, but I believe it's still better than the alternative, which is that by adding an M3 sink, you are effectively changing your metric names for _all_ configured sinks. Forcing the user into living with that behavior when you have multiple sinks of different types configured seems confusing. Doing it that way will also break any downstream aggregation/graphs that are built on top of SPIRE metrics being sent to non-M3 sinks once an M3 sink is added.

Reply:
@azdagron I think that's a reasonable compromise which would avoid confusion about potentially extraneous config parameters for most users and keep us in alignment with the flexibility of having M3 and other sinks configured at the same time. If we're all on board with that general approach, I'll revamp this in a commit by removing `EnableTypePrefix`, related documentation, and adding a multiplexer to several `metrics.Metrics` implementations in `MetricsImpl`.

@evan2645 Also wanted to understand what your concerns were with respect to reliability with @azdagron 's suggested approach of fanning out to multiple `metrics.Metrics` implementations. I couldn't think of any but just wanted to make sure there's nothing I'm missing that might bite us later.

Reply:
> If we're all on board with that general approach

Sounds fine to me so long as everything is considered (see below)

> @evan2645 Also wanted to understand what your concerns were with respect to reliability with @azdagron 's suggested approach of fanning out to multiple metrics.Metrics implementations.

Nothing in particular, I am just not sure if there are any special considerations or optimizations inside go-metrics to handle things like slow sink response, etc. Should probably have a look and see how they've done it, may be some prior art we can learn from

Reply:
One side effect I can think of is that each `metrics.Metrics` implementation will be doing stop-the-world requests of the runtime to gather runtime stats once a second (default `ProfileInterval`). Probably not a big deal though.

Reply:
There is also some duplication of adding the hostname and service name into the labels and key, respectively. Those operations would need to be done for each sink type.

There's also a check for validity of the metric key, which would be done for each sink type, but is effectively a no-op for us since we aren't configuring any allowed/blocked labels or prefixes:
https://github.com/armon/go-metrics/blob/ec5e00d3c878b2a97bbe0884ef45ffd1b4f669f5/metrics.go#L211

Other than that and what @azdagron mentioned, I didn't see any specific optimizations being done by go-metrics. The metric emission methods in `metrics.Metrics` are otherwise a passthrough to the sink implementations.

Another thing is that I don't foresee that having more than one sink configured, let alone more than two, as being a very common use case. So I don't know if we should focus too much on optimizing for it.

Reply:
> Another thing is that I don't foresee that having more than one sink configured, let alone more than two, as being a very common use case. So I don't know if we should focus too much on optimizing for it.

I agree. If the runtime collection every shows up in a profile, we can disable runtime collection on each go-metrics instance, crib the collection code, and start our own goroutine that collects once and pushes into each instance.